src/lib.rs

```diff
@@ -887,7 +903,43 @@ pub unsafe extern "C" fn path_open(
     fdflags: Fdflags,
     opened_fd: *mut Fd,
 ) -> Errno {
-    unreachable()
+    drop(fs_rights_inheriting);
+
+    let path = slice::from_raw_parts(path_ptr, path_len);
+    let at_flags = at_flags_from_lookupflags(dirflags);
+    let o_flags = o_flags_from_oflags(oflags);
+    let flags = flags_from_descriptor_flags(fs_rights_base, fdflags);
+    let mode = wasi_filesystem::Mode::READABLE | wasi_filesystem::Mode::WRITEABLE;
+
+    State::with_mut(|state| {
+        let desc = state.get_mut(fd)?;
+
+        let file = state.get_dir(fd)?;
+        let result = wasi_filesystem::open_at(file.fd, at_flags, path, o_flags, flags, mode)?;
+        let desc = Descriptor::File(File {
+            fd: result,
+            position: Cell::new(0),
+        });
+
+        let fd = match state.closed {
+            // No free fds; create a new one.
+            None => state.push_desc(desc)?,
```

Comment:
For completeness I think this'll want to close the `result` if an error happens.

Reply:
The error code from `push_desc` may also want to be changed to something more appropriate along the lines of `ENOMEM` or something to indicate that space ran out of an otherwise valid operation.